dcplib/etl/__init__.py

```diff
@@ -60,33 +53,50 @@ def dss_client(self):
-        if not os.path.exists(f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}"):
-            logger.debug("Linking fetched file %s/%s", bundle_uuid, f["name"])
-            os.symlink(f"../../files/{f['uuid']}.{f['version']}",
-                       f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}")
+    def extract_one(self, bundle_uuid, bundle_version, transformer: callable = None):
+        bundle_uuid, bundle_version, fetched_files = self.get_files_to_fetch_for_bundle(bundle_uuid, bundle_version)
+        bundle_path = f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}"
+        bundle_manifest_path = f"{self.sd}/bundle_manifests/{bundle_uuid}.{bundle_version}.json"
+        if not os.path.exists(bundle_path):
+            if self._dispatch_on_empty_bundles:
+                bundle_path = None
+            else:
+                return
+        if transformer is not None:
+            tb = transformer(bundle_uuid=bundle_uuid, bundle_version=bundle_version, bundle_path=bundle_path,
+                             bundle_manifest_path=bundle_manifest_path, extractor=self)
 
-    def get_file(self, f, bundle_uuid, bundle_version, print_progress=True):
-        logger.debug("[%s] Fetching %s:%s", threading.current_thread().getName(), bundle_uuid, f["name"])
-        res = http.get(f"{self.dss_client.host}/files/{f['uuid']}", params={"replica": "aws", "version": f["version"]})
-        res.raise_for_status()
-        with open(f"{self.sd}/files/{f['uuid']}.{f['version']}", "wb") as fh:
-            fh.write(res.content)
-        self.link_file(bundle_uuid, bundle_version, f)
-        logger.debug("Wrote %s:%s", bundle_uuid, f["name"])
-        if print_progress:
-            sys.stdout.write(".")
-            sys.stdout.flush()
-        return f, bundle_uuid, bundle_version
+            return tb
 
-    def should_fetch_file(self, f):
-        if any(fnmatchcase(f["content-type"], p) for p in self.content_type_patterns):
-            return True
-        if any(fnmatchcase(f["name"], p) for p in self.filename_patterns):
-            return True
-        return False
+    def extract(self, query, max_workers=512, max_dispatchers=1,
+                dispatch_executor_class: concurrent.futures.Executor = concurrent.futures.ThreadPoolExecutor,
+                transformer: callable = None, loader: callable = None, finalizer: callable = None):
+        futures = []
+        if query is None:
+            query = self.default_bundle_query
+        total_bundles = self.dss_client.post_search(es_query=query, replica="aws")["total_hits"]
+        if total_bundles == 0:
+            logger.error("No bundles found, nothing to do")
+            return
+        logger.info("Scanning %s bundles", total_bundles)
+        bundles = []
+        with dispatch_executor_class(max_workers=max_workers) as executor:
+            for b in self.dss_client.post_search.iterate(es_query=query, replica="aws", per_page=500):
```

Comment:
As discussed in person: to interleave ET and L, it's probably best to switch this from `iterate()` to `paginate()` and feed the resulting bundles to loader 500 at a time.

Also, while I think we should preserve the signature of `extract()`, we could try something else to enable batch loading:
- Consider adding a new callback parameter, `multi_loader: callable = None`, which can be specified mutually exclusive with `loader`, and gets called with an array of transformed bundles instead of just one.

Reply:
Here is another pattern to think about: add `__iter__` to turn `DSSExtractor` into a generator. Then refactor transform and load to subclass DSSExtractor by overloading generators too:
```
class DCPQueryTransformer(DSSExtractor):
    def transform(self, bundle):
        return transformed_bundle

    def __iter__(self):
        for bundle in DSSExtractor(extractor_opts):
            yield from map(self.transform, super().__iter__())

class DCPQueryLoader(DCPQueryTransformer):
    db_load_batch_size=100

    def load_many(self, bundles):
        ...

    def __iter__(self):
        for batch in zip(*([super().__iter__()] * db_load_batch_size)):
            self.load_many(batch)
```